text/0027-filter-and-facet-behavior.md

```diff
@@ -0,0 +1,237 @@
+- Title: Filter and Facet Behavior
+- Specification PR: [#27](https://github.com/meilisearch/specifications/pull/27)
+- MeiliSearch Issues: [#70](https://github.com/meilisearch/transplant/issues/70), [#81](https://github.com/meilisearch/transplant/issues/81), [#140](https://github.com/meilisearch/transplant/issues/140), [#152](https://github.com/meilisearch/milli/issues/152)
+
+# Facet and Filter Behavior
+
+## 1. Functional Specification
+
+### I. Summary
+
+Contrary to MeiliSearch v0.20.0, in the search engine that will be released in the v0.21.0, the users need to declare in `attributesForFaceting` the attributes on which they want to apply the filters.
+
+Why? Under the hood and during indexing phase, the core engine creates structures of data that improve the performances drastically during a search with filters.
+
+### II. Motivation
+
+Because the users need to set the attributes to `attributesForFaceting` no matter what they are using (`filters` or `facetFilters`) during the search, we need to re-define the usage of the filters/facets and stay as ISO as possible with the MeiliSearch v0.20.0.
+
+### III. Additional Materials
+
+N/A
+
+### IV. Explanation
+
+#### Remove `facetFilters`, keep `filters`
+
+The usage of `facetsFilter` is not needed anymore since everything is doable by only using the `filters` parameter.
+
+```json
+// Settings
+{
+    "attributesForFaceting": ["author"]
+}
+// Search
+{
+    "q": "",
+    "filters": "price < 20",
+    "facetFilters": ["author:'JK Rowling'"]
+}
+```
+
+becomes
+
+```json
+// Settings
+{
+    "attributesForFaceting": ["author", "price"]
+}
+// Search
+{
+    "q": "",
+    "filters": "price < 20 AND author = 'JK Rowling'"
+}
+```
+
+We decided to replace the `:` operator in favor of `=`.
+
+#### Operator behavior during the search
+
+During the search, the operators should behave as they already do in MeiliSearch v0.20.0.
+
+Here is a quick reminder of the v0.20.0 operator behavior:
+
+- `<`, `=<`, `>`, `=>` => only operates on numbers. MeiliSearch returns only the documents with numbers for this field.
+ex: `price > 19` does not return `"price": "20"` but returns `"price": 20` -> no field value conversions is done.
+- `=`, `!=`/`NOT` => string and number operation. MeiliSearch returns the documents with numbers, string (and array of strings) for this field.
+- cannot filter on `null`, objects, arrays of "undefined elements" (ex: array of `null`)
+
+#### Accepted syntaxes for `filters`
+
+Two syntaxes will be accepted for the `filters` parameter during the search:
+
+- the string syntax with the `AND`/`OR`/`NOT` operators combined with parentheses
+ex:
+```json
+{
+    "filters": "(genres = Comedy OR genres = Romance) AND director = 'Mati Diop'"
+}
+```
+
+- the array syntax:
```

Comment:
@gmourier
Checked with @Kerollmops, one question: can we accept the string syntax in an element of the array syntax? Kero says it's technically possible 🙂 ex: `filters: [["price = 20 OR size < 12"], "NOT genre = comedy"]`

Reply:
> I don't know if it's too late to suggest this, but I would strongly recommend we **change the `filters` parameter to just `filter`**. I think it's less confusing, mainly because the parameter's value will only ever be a single filter string or array.
> 
> Even if the value can be nested allowing for complexity, it's still just a logical expression. It's not as though you can define multiple filters that are applied independently or sequentially.
> 
> I would also recommend `filterDistribution` and `filterAttributes` to replace `facetsDistribution` and `attributesForFaceting`.

@curquiza Can we rename `filters` to `filter` at this time? I also think it helps to use it this way as a contextual verb.

@react-learner I think we need to keep `facetsDistribution` as is because this field is used to return computed facets. The parameter name is contextually ok. On another side, I agree with you on `attributesForFaceting` since it's now used to enable filtering on those listed fields and also to ask for MeiliSearch to expose those fields in `facetsDistribution`.